### ISSUE TITLE
Proposing a "computed-property-as-needed" rule

### ISSUE THREAD

**Issue by tlhunter** [eslint/eslint#5402]
With ES6 Computed Properties, one can do the following:

```javascript
let prop = 'b';
return {
  'a': 1,
  [prop]: 2,
  c: 3,
  ['d']: 4
}
```

Note how the brackets around the `['d']` property are superfluous. It could instead be written as `'d'` or `d`.

**When does this rule warn? Please describe and show example code:**

If enabled, the following would trigger:

```javascript
{
  ['x']: 1
}
```

Any property which does not need to be computed could trigger the rule.

**Is this rule preventing an error or is it stylistic?**

Stylistic

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

This rule affects basic ES6 syntax for situations where there are two equivalent ways to write the same code.

**Are you willing to create the rule yourself?**

I will attempt to if nobody else is interested.

**Comment by nzakas** [comment]
Thanks for the suggestion. I'm not sure this makes sense as a standalone rule. If anything, it seems like maybe it would be an option on [`quote-props`](http://eslint.org/docs/rules/quote-props.html).

**Comment by tlhunter** [comment]
You may be right, @nzakas. Perhaps the "as-needed" could throw a warning if superfluous computed property brackets are added.

The "consistent" and "consistent-as-needed" rules may be weird though, surely if one property should be computed, we wouldn't want to require all of them to be? I'm a fan of "consistent-as-needed" myself but wouldn't want all properties to need brackets if only one is computed.

**Comment by nzakas** [comment]
We probably need an option that opts-in to such a check, but otherwise :+1: . Do you want to submit a PR?

**Comment by kaicataldo** [comment]
I can try to take a look at this

**Comment by kaicataldo** [comment]
Sorry, haven't been able to get around to this. Others should feel free to take it on if they're willing and able.

**Comment by BYK** [comment]
I'm very close to submitting a PR, working on the docs.

**Comment by nzakas** [comment]
Looking at this again, I think I may be wrong about bundling this functionality with `quote-props`. From @BYK's PR, it just doesn't seem to fit into the rule that well.

I think we need to go back to the drawing board on this and decide if it makes sense to have a separate core rule or not. I'm leaning towards not at the moment, and it would be better as a custom rule, but I'd like to hear what others think.

**Comment by BYK** [comment]
May be we just need a new rule with a better organization of options: `property-style`?

**Comment by BYK** [comment]
Or `object-literal-properties`, `property-definition-style` etc.

**Comment by mysticatea** [comment]
Or just `no-useless-computed-property`. That rule would be very similar to `dot-notation` rule.

**Comment by nzakas** [comment]
Maybe `no-useless-computed-key`? We tend to use `key` for object literals such as `no-dupe-keys` and `key-spacing `.

**Comment by kaicataldo** [comment]
Using `key` makes sense to me - it's more specific and more accurately describes what the rule is checking. I'm personally not a huge fan of `useless`, though I think it works. I'll throw in my suggestion: `no-unneccessary-computed-key`?

**Comment by nzakas** [comment]
@kaicataldo we already have several rules that use `useless`, so best to stick with it.

**Comment by mysticatea** [comment]
`no-useless-computed-key` is good to me.

**Comment by BYK** [comment]
Looks like we are converging on `no-useless-computed-key`. Should I update my PR accordingly?

**Comment by nzakas** [comment]
Yup

**Comment by tlhunter** [comment]
Thanks, @BYK!
